### ISSUE TITLE
Api Utilities helper module/crate

### ISSUE THREAD

**Issue by clux** [clux/kube-rs#428]
Lots of use cases for adding helpers when dealing with common interfaces like:

- finalizers #291
- conditions #427
- annotations
- timestamps
- ownerrefs

This issue is an umbrella issue to cross-link `utils`, plus a place to discuss a possible feature'd module in kube, or a kube-utils crate.

**Comment by clux** [comment]
A couple of other helpers that I think would be useful:

- `MetadataExt` trait implementable on `k8s_openapi::ObjectMeta` (getters/setters)
- `ObjectExt` trait for `DynamicObject` (since it's the blessed dynamic type), maybe `Object`, and other `k8s_openapi::Resource` implementors (getters setters straight into metadata, typemeta) - [inspiration](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions)

**Comment by MatteoJoliveau** [comment]
Hi @clux, reviving this old issue to inquire regarding its status, specifically about helpers for owner refs. 

I've been porting a subset of [controllerutils](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil) to Rust to implement functions like `set_controller_reference` and `set_owner_reference` in a project and would love to donate the code to kube-rs. Would you be interested in a PR?

**Comment by clux** [comment]
Yeah, helpers like that would be very much appreciated! I think we have some logic like this already [inside the finalizer helper](https://github.com/kube-rs/kube-rs/blob/0e70871e10160f7ffa0c19c6003202e942537ff1/kube-runtime/src/finalizer.rs#L154-L172) that is utilizing jsonpatches. I personally always wanted that stuff to be a bit easier, at least for standard stuff, so implementations / sketches / ideas here here would be welcome.

Thinking this stuff would probably live in the core crate (if it's using jsonpatch), and then re-used down the line (e.g. the finalizer helper). If it's in core i am sure it would attract a bit of bikeshedding though :D

**Comment by MatteoJoliveau** [comment]
My helpers do not interact with any API by themselves. They're a straight port of controllerutil and only modify the Resource.

For instance here is the signature for `set_controller_reference`

```rust
fn set_controller_reference<O: Resource, K: Resource>(owner: &O, controlled: &mut K,) -> Result<(), AlreadyOwnedError>
```

**Comment by clux** [comment]
Ah, that type of module would also be work inside core, not what I expected, but it's probably easier than combining json patches, and it's a clever signature.

**Comment by teozkr** [comment]
Yeah.. I think finalizers are a kind of special case, since they have "fun" timing requirements (the finalizer MUST be acknowledged by K8s BEFORE we start performing work that they'd clean up, and so on). Most should be fine to just integrate with the regular batching.

**Comment by johanfleury** [comment]
It would be nice to have ports of `controllerutil.ContainsFinalizer(obj, finalizerName)` and `AddFinalizer(obj, finalizerName)`.
